src/hdmf_zarr/backend.py

```diff
@@ -1018,11 +1018,20 @@ def write_dataset(self, **kwargs):  # noqa: C901
                                               **options['io_settings'])
                 self._written_builders.set_written(builder)  # record that the builder has been written
                 dset.attrs['zarr_dtype'] = type_str
+                new_items = []
                 for j, item in enumerate(data):
                     new_item = list(item)
                     for i in refs:
                         new_item[i] = self.__get_ref(item[i], export_source=export_source)
-                    dset[j] = new_item
+                    new_items.append(tuple(new_item))
+                    # dset[j] = new_item
+
+                generated_dtype = []
+                for item in builder.dtype:
+                    if item['dtype'] == type('string'):
+                        item['dtype'] = 'U25'
```

Comment:
this mutates the builder's dtype right?

Reply:
Yes. When numpy gets the "str class" it removes the actual value and on read returns an empty string.

Reply:
not sure what you mean, when numpy gets `str` where? What i mean is that this would mutate the `builder.dtype` dict, which is almost always undesirable, especially since the builder is passed in as an arg, mutating your args will buy you invisible bugs later.

eg.

```python
>>> mydtypes = [{'dtype': 'int', 'name': 'test'}, {'dtype': 'float', 'name': 'test2'}]
>>> for item in mydtypes:
...    if item['dtype'] == 'int':
...        item['dtype'] = 'newvalue'
>>> mydtypes
[{'dtype': 'newvalue', 'name': 'test'}, {'dtype': 'float', 'name': 'test2'}]
```

instead do 

```python
for item in builder.dtype:
    if item['dtype'] is str:
        generated_dtype.append((item['name'], 'U25'))
    else:
        generated_dtype.append((item['name'], item['dtype']))
```

which doesn't mutate `builder.dtype`. see https://github.com/hdmf-dev/hdmf-zarr/blob/8100341512324af25718ab93dc7fc54b68f03469/src/hdmf_zarr/backend.py#L1028-L1036

Reply:
I see what you did.